Refactor changelog to use new `\ChangelogRef`

BryantLam opened the pull request:
This PR primarily refactors the references in the changelog entries to reduce inconsistencies between entries and to make it easier to later alter its format if desired.

**I highly recommend reading the changes in each commit rather than all changes together.** This is because https://github.com/jdinan/specification/commit/eff02647eee3838074974a65e624b96bb86854f1 adds new deprecation-specific labels and then applies them into the changelog &mdash; in my opinion this is the only non-trivial change to review and I don't want it to get muddied by the mostly mechanical conversion done in https://github.com/jdinan/specification/commit/e3dd9d170cbf22693a713b1e71b056aee8bd1384 which are the bulk of the changes.

Notes:
1. https://github.com/jdinan/specification/commit/eff02647eee3838074974a65e624b96bb86854f1 tries to add references to each deprecated item. Turns out some of the older changelog entries were somewhat ambiguously phrased, so these references should definitely help disambiguate the changes in each of the affected changelog entries.
2. https://github.com/jdinan/specification/commit/e3dd9d170cbf22693a713b1e71b056aee8bd1384
    - Each `\ChangelogRef` ends in a `%` sigil as a visual reminder to not add a period.
    - As a restriction, I didn't reword any of the existing entries to make this commit as mechanical as possible.
3. `\ChangelogRef` is implemented using package `cleveref`. By default, references are `sort&compress`. For our purposes, sort is fine. Compress, however, means that many-reference entries are compressed into a range instead; e.g., `See Sections 1, 2, 3, 4 and 5.` is converted into `See Sections 1 to 5.`. **If this behavior is not desired, I can easily change it.**
    - There's a bypass. One of the changelog entries uses double-commas `,,` because it references tables and tables are spread throughout the document; its references shouldn't be compressed.
4. `cleveref` doesn't use the Oxford comma (:cry:). I left it this way, but if we strongly oppose to its non-use, I can change that too.

BryantLam committed:
Make \cref always capitalize

BryantLam committed:
Changelog: Add refs for deprecated and removed items

BryantLam committed:
Changelog: Convert semi-automatically to use `\ChangelogRef`

One exceptional case involved several references to tables. This item
now uses double-commas to separate the ref entries so that `cleveref`
will not use range compression.

BryantLam committed:
Add \Crefname{paragraph} to support our custom paragraph headings

jdinan commented on content/backmatter.tex:
```diff
@@ -540,7 +549,8 @@ \subsection{\textit{C11} and \CorCpp: \FUNC{shmem\_fetch}, \FUNC{shmem\_set}, %%
 expanded for clarity to ``compare\_swap'', ``fetch\_inc'', and ``fetch\_add''.
 
-\subsection{\textit{Fortran} API}\label{subsec:deprecate-fortran} %% WARNING: Issue #66.
+\subsection{\textit{Fortran} API} %% WARNING: Issue #66.
```
Can we remove this comment? Looks like it may correspond to an issue you filed a few years ago - https://github.com/openshmem-org/specification/issues/66.

BryantLam replied:
I'd rather leave the comment. It's a visual reminder for why `\textit{Fortran}` is used instead of `\Fortran` (the PDF bookmark is broken when using commands that take optional arguments). This issue did come up once during the v1.5rc2 period.

BryantLam replied:
(To clarify, that root problem with https://github.com/openshmem-org/specification/issues/66 was not resolved. We punted on it.)

jdinan replied:
Could you add a little more context to clarify what the warning is about? Something like:
```
%% WARNING: Need to use \textit{Fortran} instead of \Fortran, see Issue #66 for details
```

BryantLam replied:
:+1: I clarified the problem at the bottom of the deprecated-rationale entries. It'll remind document editors when they add new entries to the list.

jdinan replied:
Thanks! I was absolutely thinking of myself, that I would forget why we needed to handle the Fortran case in this way. :)

jdinan commented on content/backmatter.tex:
```diff
@@ -637,20 +651,22 @@ \section{Version 1.5}
 %
 \item Deprecated active-set-based library constants and collective functions.
-\\ See Sections \ref{subsec:library_constants} and \ref{subsec:coll}.
+\ChangelogRef{dep:active_set_libconst_and_collectives, dep:shmem_barrier}%
```
This removes the references to the library constants and collectives sections and replaces them with references to the deprecation annex. I'm wondering if we want to have both references?

BryantLam replied:
~~My motivation was that&mdash;historically&mdash;many deprecated items didn't have references at all~~ and now that I'm adding them, having only a ref to the deprecation rationale made sense to me since the deprecated interfaces are listed in the rationales too. **Edit:** Now that I looked at it, I was wrong. Most did have references; the "many" I was referring to were for Fortran and SHMEM_CACHE...

Nonetheless, I'm okay with both references if we want to do it that way. I'd have to go back and review what relevant sections should be included for each deprecated item.

jdinan replied:
This is a case where the APIs are still in the document, but marked as deprecated. In these cases, I think we should reference both the deprecation annex and the API definitions. Alternatively, we could reference the API definitions from the deprecation annex, but I don't think we are doing this right now.

jdinan replied:
@BryantLam Any change needed here?

BryantLam replied:
All good. This was resolved in https://github.com/jdinan/specification/commit/f025b115809c08582c1be869cbbaa0b48271194d#diff-3b37f8686e2f1bb2e5cb250aa60d8d16R672-R676.

jdinan commented on content/backmatter.tex:
```diff
@@ -829,72 +856,72 @@ \section{Version 1.4}
 %
 \item Deprecated the \FUNC{shmem\_wait} functions and the \CTYPE{long}-typed \CorCpp \FUNC{shmem\_wait\_until} function.
-\\ See Section \ref{subsec:p2p_intro}.
+\ChangelogRef{dep:shmem_wait, dep:long_typed_shmem_wait_until}%
```
This is another case where we have lost a reference.

BryantLam replied:
I'll change this pointer to include `subsec:shmem_wait_until` instead of `subsec:p2p_intro`.

jdinan replied:
:+1:

jdinan commented on content/backmatter.tex:
```diff
@@ -916,38 +943,39 @@ \section{Version 1.3}
 %
 \item Deprecation of all constants that start with \CONST{\_SHMEM\_*}.
-\\See Section \ref{subsec:library_constants}.
+\ChangelogRef{dep:libconst_undershmem}%
```
Another lost reference.

jdinan commented on content/backmatter.tex:
```diff
@@ -972,52 +1000,52 @@ \section{Version 1.2}
 %
 \item Deprecation of \FUNC{\_my\_pe} and \FUNC{\_num\_pes} routines.
-\\See Sections \ref{subsec:shmem_my_pe} and \ref{subsec:shmem_n_pes}.
+\ChangelogRef{dep:func_not_shmemunder}%
```
Lost reference.

jdinan commented on content/backmatter.tex:
```diff
@@ -972,52 +1000,52 @@ \section{Version 1.2}
 %
 \item Deprecation of \Fortran API routine \FUNC{SHMEM\_PUT}.
-\\See Section \ref{subsec:shmem_put}.
+\ChangelogRef{dep:fortran_shmem_put}%
```
Lost reference.

BryantLam replied:
This one feels dubious now that Fortran is removed. Should it still point to `subsec:shmem_put`?

BryantLam replied:
I left it in. It's at least historically accurate.

jdinan replied:
That's a fair point. Once the interfaces are removed, the only meaningful thing to reference in the current document is the deprecation annex. What about adding a new macro for referencing across versions of the document, something like `\crossref{1.4}{9.5.1}` -> `OpenSHMEM 1.4, Section 9.5.1`?

BryantLam replied:
Resolved. I wrote it in directly instead of adding a new command because:
- I don't expect further removals now that Fortran and SHMEM_CACHE are removed.
- There's potentially a lot of variations that would be non-trivial to code for. E.g., a Sections range, the Annexes, or Tables, etc. Alas, `\cref` would be ideal if the old-doc references could be hardcoded...

jdinan commented on content/environment_variables.tex:
```diff
@@ -8,9 +8,9 @@
-\hyperref[subsec:deprecate-sma-env]{\VAR{SMA\_*} Environment Variables}
-deprecation rationale for more details.
+establishes the controlling value.
+Refer to the \ENVVAR{SMA\_*} Environment Variables deprecation rationale,
```
```suggestion
Refer to the \ENVVAR{SMA\_*} environment variables deprecation rationale,
```

jdinan reviewed (changes requested):
Looks great -- thanks for these edits! A few minor nitpicks and also wondering whether we want to add back some changelog references that were substituted with references to the deprecation annex.

BryantLam committed:
Update content/environment_variables.tex

Co-Authored-By: James Dinan <jdinan@users.noreply.github.com>

BryantLam committed:
Changelog: Restore references to main sections for deprecated items

BryantLam committed:
Apply code review comments from @jdinan to backmatter

- Clarify problem with openshmem-org#66
- Change reference for deprecated (now removed) Fortran SHMEM_PUT to
  cross-reference an OpenSHMEM 1.4 section
    - Add optional text space in `\ChangelogRef`

jdinan commented on utils/defs.tex:
```diff
@@ -135,7 +135,7 @@
 }
 \makeatother
-\newcommand{\ChangelogRef}[1]{\\See \cref{#1}.}
+\newcommand{\ChangelogRef}[1]{\\See \cref{#1}.\space}
```
Do you still need `\space` if you remove the `%` immediately after the usage of this macro?

BryantLam replied:
Nope. You don't need `\space`, but it doesn't hurt either because the space is elided if there's no subsequent sentence. I'd prefer to have the `%` visual reminder even if it causes this weird case here with the second Fortran SHMEM_PUT sentence.
